forem/forem
# Updates the email_addresses hash with the contact info

## What type of PR is this? (check all applicable)


## Description

This PR deserves its own incident retro and I'll create one accordingly but you can read more about the context of this PR below:

### Context: ###
https://dev-internal.slack.com/archives/C5BJE4PGB/p1607611327262300
It was discovered that the contact email is not showing up on DEV. The contact email was added in this PR, however since we were modifying an existing Site Config field (i.e. email_addresses), we were meant to update the value in the DB as well.

We'd never added this field to the db and hence as a result all of the references to SiteConfig[:email_addresses][:contact] were blank. The contact email was also not showing up as a field on /admin/config.

I'd optimistically implemented the above in a data script in [this PR](https://github.com/forem/forem/pull/11848). However, it seems that this script DID NOT actually update the value. 

Below you can debug with me and find out why:

### Some more Investigation on why [this PR](https://github.com/forem/forem/pull/11848) did not work: ###

Let's view the Site Config email addresses locally and try to add a contact email key as per PR 11848.
```
[18] pry(main)> SiteConfig.email_addresses
  SiteConfig Load (0.5ms)  SELECT var, value FROM "site_configs"
=> {"default"=>"yo@dev.to", "business"=>"biz@dev.to", "privacy"=>"privacy@dev.to", "members"=>"yo@dev.to"}
```

Now, let's try adding the contact email address as an example:
```
[19] pry(main)> SiteConfig.email_addresses[:contact] = "contact@dev.to"
=> "test@dev.to"
```

Let's view the email addresses again and see if this change was set in the db (note: we didn't see any create commit when we tried to set it above):
```
[20] pry(main)> SiteConfig.email_addresses
=> {"default"=>"yo@dev.to", "business"=>"biz@dev.to", "privacy"=>"privacy@dev.to", "members"=>"yo@dev.to", "contact"=>"contact@dev.to"}
```

Hmmm it looks like it has been set, but let's make sure and reload the console and check again:
```
[21] pry(main)> reload!
Reloading...
  ProfileField Load (0.6ms)  SELECT "profile_fields".* FROM "profile_fields" ORDER BY "profile_fields"."id" ASC LIMIT $1  [["LIMIT", 1000]]
  User Exists? (0.5ms)  SELECT 1 AS one FROM "users" LIMIT $1  [["LIMIT", 1]]
Unset ENV variable: COMMUNITY_COPYRIGHT_START_YEAR.
Unset ENV variable: STRIPE_SECRET_KEY.
Unset ENV variable: STRIPE_PUBLISHABLE_KEY.
  SiteConfig Load (0.3ms)  SELECT var, value FROM "site_configs"
=> true
```

Now let's double check **again** that the contact email was set:
```
[22] pry(main)> SiteConfig.email_addresses
  SiteConfig Load (0.4ms)  SELECT var, value FROM "site_configs"
=> {"default"=>"yo@dev.to", "business"=>"biz@dev.to", "privacy"=>"privacy@dev.to", "members"=>"yo@dev.to"}
```

It **was not set**!
The above summarizes what went wrong in [PR 11848](https://github.com/forem/forem/pull/11848).

### What's next: ###

This felt related to the difference between a simple object and a hash. So we (@rhymes and I) tried the following. We set the entire hash again but added the contact value
```
SiteConfig.email_addresses = {
[63] pry(main)*   default: SiteConfig.email_addresses[:default],
[63] pry(main)*   business: SiteConfig.email_addresses[:business],
[63] pry(main)*   privacy: SiteConfig.email_addresses[:privacy],
[63] pry(main)*   members: SiteConfig.email_addresses[:members],
[63] pry(main)*   contact: ApplicationConfig["DEFAULT_EMAIL"]
[63] pry(main)* }
  SiteConfig Load (1.2ms)  SELECT "site_configs".* FROM "site_configs" WHERE "site_configs"."var" = $1 LIMIT $2  [["var", "email_addresses"], ["LIMIT", 1]]
=> {:default=>"default@dev.to", :business=>"business@dev.to", :privacy=>"privacy@dev.to", :members=>"members@dev.to", :contact=>"yo@dev.to"}
```

We then tried reloading:
```
[64] pry(main)> reload!
Reloading...
  ProfileField Load (0.3ms)  SELECT "profile_fields".* FROM "profile_fields" ORDER BY "profile_fields"."id" ASC LIMIT $1  [["LIMIT", 1000]]
  User Exists? (1.2ms)  SELECT 1 AS one FROM "users" LIMIT $1  [["LIMIT", 1]]
Unset ENV variable: COMMUNITY_COPYRIGHT_START_YEAR.
Unset ENV variable: STRIPE_SECRET_KEY.
Unset ENV variable: STRIPE_PUBLISHABLE_KEY.
  SiteConfig Load (0.3ms)  SELECT var, value FROM "site_configs"
=> true
```

Then we checked again if contact was set:
```
[65] pry(main)> SiteConfig.email_addresses
  SiteConfig Load (0.4ms)  SELECT var, value FROM "site_configs"
=> {"default"=>"default@dev.to", "business"=>"business@dev.to", "privacy"=>"privacy@dev.to", "members"=>"members@dev.to", "contact"=>"yo@dev.to"}
```
It was! I honestly am not 100% clear why the[ first attempt](https://github.com/forem/forem/pull/11848) wouldn't have worked, but I do suspect that in order to trigger the attribute setter on the SiteConfig hash object, it seems like we need to set all the values in the hash and not only the one we are changing. 

This is a high priority and so I created this PR and will look into the rationale in more detail in the retro. 

## Related Tickets & Documents
https://github.com/forem/forem/issues/11959

## QA Instructions, Screenshots, Recordings

1. Open the rails console with `rails c`
2. Get the current SiteConfig.email_addresses, mine has been updated and looks like this:
```
  SiteConfig Load (0.5ms)  SELECT var, value FROM "site_configs"
=> {"default"=>"default@dev.to", "business"=>"business@dev.to", "privacy"=>"privacy@dev.to", "members"=>"members@dev.to"}
```
3. Go back to the terminal and run `rails data_updates:run`
4. Switch to the rails console again and type `reload!`. 
5. Check if the contact value has been added to the hash. You should see something like this: 
```
 SiteConfig.email_addresses
  SiteConfig Load (0.5ms)  SELECT var, value FROM "site_configs"
=> {"default"=>"default@dev.to", "contact"=>"contact@dev.to", "business"=>"business@dev.to", "privacy"=>"privacy@dev.to", "members"=>"members@dev.to"}
```
6. To double check everything, open up your /admin/config and go to the site config email section. You should see the contact field appear:
<img width="1677" alt="Screenshot 2020-12-18 at 10 45 23" src="https://user-images.githubusercontent.com/2786819/102593931-36413700-411e-11eb-8915-766f1000c70b.png">

_Please replace this line with instructions on how to test your changes, as well
as any relevant images for UI changes._

### UI accessibility concerns?

_If your PR includes UI changes, please replace this line with details on how
accessibility is impacted and tested. For more info, check out the
[Forem Accessibility Docs](https://docs.forem.com/frontend/accessibility)._

## Added tests?

      have not been included_

## Added to documentation?

      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)

## [optional] Are there any post deployment tasks we need to perform?

## [optional] What gif best describes this PR or how it makes you feel?

![alt_text](https://media.giphy.com/media/3otPosozzXEVzPjkC4/giphy.gif)
